pkg/storage/export/export/vm-source.go

```diff
@@ -0,0 +1,245 @@
+		}
+	}
+	return key
+}
+
+func (ctrl *VMExportController) getVMFromVMI(vmi *virtv1.VirtualMachineInstance) *virtv1.VirtualMachine {
+	ownerRef := metav1.GetControllerOf(vmi)
+	if ownerRef != nil {
+		if ownerRef.Kind == "VirtualMachine" && ownerRef.APIVersion == virtv1.GroupVersion.String() {
+			if vm, exists, err := ctrl.getVm(vmi.Namespace, ownerRef.Name); !exists || err != nil {
+				log.Log.V(3).Infof("Unable to get owner VM %s/%s for VMI %s/%s", vmi.Namespace, ownerRef.Name, vmi.Namespace, vmi.Name)
+			} else {
+				return vm
+			}
+		}
+	}
+	return nil
+}
+
+func (ctrl *VMExportController) isSourceAvailableVM(vmExport *exportv1.VirtualMachineExport) (bool, string, error) {
+	vmi, exists, err := ctrl.getVmi(vmExport.Namespace, vmExport.Spec.Source.Name)
+	if err != nil {
+		return false, "", err
+	}
+	if exists {
+		return false, fmt.Sprintf("Virtual Machine %s/%s is running", vmi.Namespace, vmi.Name), nil
+	}
+	return true, "", nil
+}
+
+func (ctrl *VMExportController) getPVCFromSourceVM(vmExport *exportv1.VirtualMachineExport) (*sourceVolumes, error) {
+	pvcs, allPopulated, err := ctrl.getPVCsFromVM(vmExport.Namespace, vmExport.Spec.Source.Name)
+	if err != nil {
+		return &sourceVolumes{}, err
+	}
+	log.Log.V(3).Infof("Number of volumes found for VM %s/%s, %d, allPopulated %t", vmExport.Namespace, vmExport.Spec.Source.Name, len(pvcs), allPopulated)
+	if len(pvcs) > 0 && !allPopulated {
+		return &sourceVolumes{
+			volumes:          pvcs,
+			sourceAvailable:  allPopulated,
+			availableMessage: fmt.Sprintf("Not all volumes in the Virtual Machine %s/%s are populated", vmExport.Namespace, vmExport.Spec.Source.Name)}, nil
+	}
+	sourceAvailable, availableMessage, err := ctrl.isSourceAvailableVM(vmExport)
+	if err != nil {
+		return &sourceVolumes{}, err
+	}
+	return &sourceVolumes{
+		volumes:          pvcs,
+		sourceAvailable:  sourceAvailable,
+		availableMessage: availableMessage}, nil
+}
+
+func (ctrl *VMExportController) getPVCsFromVM(vmNamespace, vmName string) ([]*corev1.PersistentVolumeClaim, bool, error) {
+	var pvcs []*corev1.PersistentVolumeClaim
+	vm, exists, err := ctrl.getVm(vmNamespace, vmName)
+	if err != nil {
+		return nil, false, err
+	}
+	if !exists {
+		return nil, false, nil
+	}
+	allPopulated := true
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		pvcName := storagetypes.PVCNameFromVirtVolume(&volume)
+		if pvcName == "" {
+			continue
+		}
+		pvc, exists, err := ctrl.getPvc(vmNamespace, pvcName)
+		if err != nil {
+			return nil, false, nil
+		}
+		if exists {
+			populated, err := ctrl.isPVCPopulated(pvc)
+			if err != nil {
+				return nil, false, err
+			}
+			pvcs = append(pvcs, pvc)
+			if !populated {
+				allPopulated = false
+			}
+		}
+	}
+	return pvcs, allPopulated, nil
+}
+
+func (ctrl *VMExportController) updateVMExportVMStatus(vmExport *exportv1.VirtualMachineExport, exporterPod *corev1.Pod, service *corev1.Service, sourceVolumes *sourceVolumes) (time.Duration, error) {
```

Comment:
I believe this function should return a requeue duration when pvcs are populating

Reply:
You are correct, fixing that.